File: packages/algo-go/solutions/algo_03.go
package solutions

func MaxSubstr(input string) int {
	max := 0
	start := 0
	indexes := [128]int{}
	for i := 0; i < 128; i++ {
		indexes[i] = -1
	}
	for i, c := range input {
		if indexes[byte(c)] >= start {
			start = indexes[byte(c)] + 1
		}
		indexes[uint8(c)] = i
		if i-start+1 > max {
			max = i - start + 1
		}
	}

	return max
}

File: packages/algo-py/solutions/algo_03.py
def max_substr(input):
    start = 0
    max_count = 0
    indexes = [-1 for _ in range(128)]
    for i in range(len(input)):
        c = ord(input[i])
        if indexes[c] >= start:
            start = indexes[c] + 1
        indexes[c] = i
        max_count = max(max_count, i - start + 1)
    return max_count
